web/frontend/src/components/slam.vue

```diff
@@ -1,193 +1,208 @@
+    getPointCloudMapStream.on('data', (res) => {
+      const chunk = res.getPointCloudPcdChunk_asU8();
+      chunks.push(chunk);
+    });
+    getPointCloudMapStream.on('status', (status) => {
+      if (status.code !== 0) {
+        const error = {
+          message: status.details,
+          code: status.code,
+          metadata: status.metadata,
+        };
+        reject(error);
+      }
+    });
+    getPointCloudMapStream.on('end', (end) => {
+      if (end === undefined || end.code !== 0) {
+        // the error will be logged in the 'status' callback
+        return;
+      }
+      const arr = concatArrayU8(chunks);
+      resolve(arr);
+    });
+  });
+};
+
+const fetchSLAMPose = (name: string): Promise<commonApi.Pose> => {
+  return new Promise((resolve, reject): void => {
+    const req = new slamApi.GetPositionNewRequest();
+    req.setName(name);
+    props.client.slamService.getPositionNew(req, new grpc.Metadata(), (error, res): void => {
+      if (error) {
+        reject(error);
+        return;
+      }
+      resolve(res!.getPose()!);
+    });
+  });
+};
 
 const refresh2d = async (name: string) => {
+  const map = await fetchSLAMMap(name);
+  const returnedPose = await fetchSLAMPose(name);
+  return Promise.all([map, returnedPose]);
+};
 
-  /*
-   * const mapPromise = fetchSLAMMap(name);
-   * const posePromise = fetchSLAMPose(name);
-   * await mapPromise;
-   * await posePromise;
-   */
-  await fetchSLAMImageMap(name);
+const scheduleRefresh2d = (name: string, time: string) => {
+  const timeoutCallback = async () => {
+    try {
+      const results = await refresh2d(name);
+      if (results && results.length === 2) {
+        [pointcloud, pose] = results;
+        pointCloudUpdateCount += 1;
+      }
+    } catch (error) {
+      displayError(error);
+      return;
+    }
+    if (refresh2DCancelled) {
+      return;
+    }
+    scheduleRefresh2d(name, time);
+  };
+  slam2dTimeoutId = window.setTimeout(timeoutCallback, Number.parseFloat(time) * 1000);
+};
+
+const scheduleRefresh3d = (name: string, time: string) => {
+  const timeoutCallback = async () => {
+    try {
+      const result = await fetchSLAMMap(name);
+      if (result) {
+        pointcloud = result;
+        pointCloudUpdateCount += 1;
+      }
+    } catch (error) {
+      displayError(error);
+      return;
+    }
+    if (refresh3DCancelled) {
+      return;
+    }
+    scheduleRefresh3d(name, time);
+  };
+  slam3dTimeoutId = window.setTimeout(timeoutCallback, Number.parseFloat(time) * 1000);
 };
 
-// eslint-disable-next-line require-await
 const updateSLAM2dRefreshFrequency = async (name: string, time: 'manual' | 'off' | string) => {
-  window.clearInterval(slam2dIntervalId);
+  refresh2DCancelled = true;
+  window.clearTimeout(slam2dTimeoutId);
 
   if (time === 'manual') {
-    refresh2d(name);
-
+    try {
+      const results = await refresh2d(name);
```

Comment:
@micheal-parks do you have a recommendation on how I can do this any cleaner?

Reply:
Anything in particular you're worried about? In general this looks fine and can be cleaned up later.

Reply:
Never mind, I figured it out.